Add deferrable mode to Kubernetes and GKE operators

MrGeorgeOwl opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

MrGeorgeOwl committed:
Add deferrable mode to Kubernetes and GKE operator

MrGeorgeOwl committed:
Fix import of the dag

VladaZakharova committed:
update kubernetesPodOperator to return xcom value

MrGeorgeOwl committed:
Change cleanup method to first implementation

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -272,6 +276,8 @@ def __init__(
         termination_grace_period: int | None = None,
         configmaps: list[str] | None = None,
+        deferrable: bool = False,
+        poll_interval: float = 2,
```
What does `poll_interval` mean for non-async operator?

MrGeorgeOwl replied:
For non-async nothing but it will give user opportunity to configure how often trigger should check pod operator status if deferrable mode is on

kaxil replied:
Please update docstring accordingly on L212 to say it is only for async

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -437,6 +442,10 @@ def hook(self) -> KubernetesHook:
     def client(self) -> CoreV1Api:
         return self.hook.core_v1_client
 
+    def get_hook(self):
+        warnings.warn("get_hook is deprecated. Please use hook instead.", DeprecationWarning, stacklevel=2)
+        return self.hook
+
```
Why this unrelated change of moving the hook position?

MrGeorgeOwl replied:
It is little refactoring for improving code readability so all cached methods and hook properties can be found in the same place

kaxil replied:
Let's avoid those unrelated changes please -- makes it difficult for checking git history

VladaZakharova replied:
We have moved this method to be placed in one place as other methods after all the cached_properties. We will avoid such kind of changes in next PRs, thanks!

kaxil commented on docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst:
```diff
@@ -131,6 +131,14 @@ And then use it in other operators:
     :start-after: [START howto_operator_gke_xcom_result]
     :end-before: [END howto_operator_gke_xcom_result]
 
+You can use deferrable mode for this action in order to run the operator asynchronously:
```
There are some limitations of using async KPO in its current state?

What are they? let's document it

MrGeorgeOwl replied:
Could you be more descriptive about the idea you propose? I am not sure what you meant. That string is made for docs reader to inform them that deferrable mode is implemented for that pattern of usage

VladaZakharova committed:
Add logic to convert config_file content to prevent memory leak

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -484,15 +496,21 @@ def extract_xcom(self, pod: k8s.V1Pod):
+        if self.deferrable:
+            self.asynchronously(context)
+        else:
+            return self.synchronously(context)
```
Can you rename this too.

@VladaZakharova Apologies about delay in our contribution - Could you rebase once https://github.com/apache/airflow/pull/28523 is merged. That PR also has additional feature of ping-pong to show periodic logs which our customers found very useful

kaxil replied:
If you prefer you contribute first and @phanikumv rebases his PR -- that is fine too

kaxil replied:
https://github.com/apache/airflow/pull/28523#issuecomment-1362176160 -- Let's get this PR merged first since it was created first. We will rebase that PR on main

VladaZakharova replied:
Sure, would be nice of you. We have added some changes to prevent memory leak and are working now on other changes to improve performance for the operator. We will let you know when it will be ready :)

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -509,48 +527,141 @@ def execute(self, context: Context):
-        with _optionally_suppress(reraise=reraise):
-            for event in self.pod_manager.read_pod_events(pod).items:
-                self.log.error("Pod Event: %s - %s", event.reason, event.message)
+    def asynchronously(self, context: Context):
```
`execute_async` would be a better name.

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -406,3 +418,83 @@ def _get_bool(val) -> bool | None:
+    async def _load_config(self):
+        """Returns Kubernetes API session for use with requests"""
+        in_cluster = self._coalesce_param(self.in_cluster, self._get_field("in_cluster"))
+        cluster_context = self._coalesce_param(self.cluster_context, self._get_field("cluster_context"))
```
`self._get_field` calls `self.conn_extras` which has a DB call. 

Can you override it please so that we can make it an async call

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -484,15 +498,21 @@ def extract_xcom(self, pod: k8s.V1Pod):
+        else:
+            return self.synchronously(context)
+
+    def synchronously(self, context: Context):
```
same as https://github.com/apache/airflow/pull/28230/files#r1054839109

`synchronously` name for a method alone doesn't make much sense

kaxil reviewed (commented):
If you can separate this into 2 PRs - one for async mode for KPO and the other for async mode for GKE, that will help in reviewing

VladaZakharova committed:
Add caching for Kubernetes client

MrGeorgeOwl committed:
Middlework.Change call to db, refactor temporary file creation and little refactoring

MrGeorgeOwl committed:
Middlework.Add caching to async request to db, remove error imports, remove method for writing permanent config file

MrGeorgeOwl committed:
Change the way of authentication to GKE cluster

MrGeorgeOwl committed:
Fix marking task failed, pending for additional resources, add failed event handling

MrGeorgeOwl committed:
Add additional check for timeout event

VladaZakharova committed:
Add logging on all stages of pod run

Add logging on all stages of pod run, also description for parameters and unit tests

VladaZakharova committed:
Fix tests for gke operator and kubernetes hook

VladaZakharova committed:
Add asgiref module to cncf provider.yaml

kaxil commented on generated/provider_dependencies.json:
```diff
@@ -209,8 +209,10 @@
       "cryptography>=2.0.0",
-      "kubernetes>=21.7.0,<24"
+      "kubernetes>=21.7.0,<24",
+      "kubernetes_asyncio==24.2.2"
```
Let's not pin to a specific version, Airflow is also used as a library and an application. Pinning it to a specific version restricts what deps can be used with Airflow

MrGeorgeOwl replied:
Change it to range

kaxil commented on airflow/providers/google/cloud/operators/kubernetes_engine.py:
```diff
@@ -305,11 +303,9 @@ def __init__(
         *,
         location: str,
         cluster_name: str,
-        use_internal_ip: bool = False,
```
This is going to break backwards compat!

MrGeorgeOwl replied:
Replace it with optional parameter and add deprecation message

kaxil commented on airflow/providers/google/cloud/operators/kubernetes_engine.py:
```diff
@@ -305,11 +303,9 @@ def __init__(
         project_id: str | None = None,
         gcp_conn_id: str = "google_cloud_default",
         impersonation_chain: str | Sequence[str] | None = None,
-        regional: bool = False,
```
This is going to break backwards compat!

MrGeorgeOwl replied:
Replace it with optional parameter and add deprecation message

MrGeorgeOwl committed:
Change version ranging, add deprecation messages

MrGeorgeOwl committed:
Add tests for GKE pod hook, refactor test for GKE operator

VladaZakharova committed:
Add tests for GKE trigger + add some params for it for correct inheritance

VladaZakharova committed:
Fix for passing static checks
